recipes/sml/all/conanfile.py

```diff
@@ -0,0 +1,37 @@
+from conans import ConanFile, tools
+from conans.tools import check_min_cppstd
+import os
+
+
+class SMLConan(ConanFile):
+    name = "sml"
+    homepage = "https://github.com/boost-ext/sml"
+    description = "SML: C++14 State Machine Library"
+    topics = ("state-machine", "boost", "metaprogramming", "design-patterns", "sml")
+    license = "BSL-1.0"
+    url = "https://github.com/conan-io/conan-center-index"
+    settings = "compiler"
+    no_copy_source = True
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def validate(self):
+        check_min_cppstd(self, "14")
```

Comment:
It prevents to consume sml if default standard of the compiler is not 14 or above, and `compiler.cppstd` not explictly set. I think that it also raises if compiler is not mainstream.
You can use this kind of pattern instead:
https://github.com/conan-io/conan-center-index/blob/d7e0581908730a3ccbe2ce397be28655b4409341/recipes/capnproto/all/conanfile.py#L41-L48
https://github.com/conan-io/conan-center-index/blob/d7e0581908730a3ccbe2ce397be28655b4409341/recipes/capnproto/all/conanfile.py#L57-L63

Reply:
I am not sure how it helps here? The only issue with my usage of `check_min_cppstd(self, "14")` is that it might not work for some strange niche compiler and the user will be forced to set cppstd for it explicitly. This eventually makes it work for any compiler.

Adding the rest of the suggested code does not help here as `tools.check_min_cppstd(self, 14)` will throw an Exception for an unknown compiler anyway so the rest of the code will not be executed, right?

Reply:
It allows to use the recipe even if cppstd (which is experimental) is not set. If you look at https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/4591/2/sml/1.1.3//summary.json, it throws for all apple-clang versions tested in CI.

Reply:
I do not think that we should verify the C++ standard version based on the compiler's version. If this is the case we need a better utility in Conan tools. It is just too verbose and error-prone.

Reply:
But if you prefer, I can rename `validate` to `configure` (this change will probably be reverted for Conan 2.0 later on) and add `if self.settings.compiler.get_safe("cppstd"):`.

Reply:
So it seems that adding `get_safe()` made the things worse. I think that leaving just `check_min_cppstd(self, "14")` is the right solution.

According to the Conan docs:

> Validates if the applied cppstd setting (from compiler.cppstd settings or deducing the default from compiler and compiler.version) is at least the value specified in the cppstd argument. It raises a ConanInvalidConfiguration when is not supported.

> If no compiler has been specified or the compiler is unknown, it raises a ConanException exception.

So if there is no `cppstd` the Conan tries to determine the version by itself. This provides a better logic than handling it by myself with `_minimum_compilers_version` suggested above.

The only difference is that there will be an exception for an unsupported compiler. In such a case we should not guess that it supports C++14 because it might not. The user should either add `cppstd` explicitly to its `settings.yml` or propose a change to Conan that will add support for a specific compiler.

Reply:
Not sure for conan 2.0, `configure()` allows early checks (before dependency graph resolution of current recipe).

You can't just add `if self.settings.compiler.get_safe("cppstd")` without all the others verifications of compiler versions (otherwise CI of CCI will try to build with lowest compiler version and it will fail).

Reply:
Learned this the hard way ;-) I think that now everything is fine?

Reply:
For my tracking purposes conan-io/conan#8002

Reply:
Please follow the suggestions above, this is a requirement because of a technical limitation of CCI... so for now we need to make sure this is merged in.

Thanks for understanding 😄

Reply:
Sure, I changed to `configure()` usage 4 days ago.

Reply:
Yes but you have to implement the full solution proposed above (with explicit compiler versions).